include/fiction/io/print_layout.hpp

```diff
@@ -307,6 +316,86 @@ void print_cell_level_layout(std::ostream& os, const Lyt& layout, const bool io_
     // flush stream
     os << std::endl;
 }
+/**
+ * Writes a simplified 2D representation of a SiDB charge layout to an output stream.
+ *
+ * @tparam Lyt Cell-level layout based in SiQAD-coordinates.
+ * @param os Output stream to write into.
+ * @param cds The charge distribution surface of which the charge layout is to be printed.
+ * @param cs_color Flag to utilize color escapes for charge states.
+ */
+template <typename Lyt>
+void print_charge_layout(std::ostream& os, const charge_distribution_surface<Lyt>& cds,
+                         const bool cs_color = false) noexcept
+{
+    // empty layout
+    if (cds.num_cells() == 0ul)
+    {
+        os << "[i] empty layout" << std::endl;
+        return;
+    }
+
+    for (decltype(cds.y()) y_pos = 0; y_pos <= cds.y(); ++y_pos)
+    {
+        for (decltype(cds.x()) x_pos = 0; x_pos <= cds.x(); ++x_pos)
+        {
+            cell<Lyt> c{x_pos, y_pos};
+
+            const fmt::text_style no_color{};
+
+            const auto ct = cds.get_cell_type(c);
+
+            if (Lyt::technology::is_normal_cell(ct))
+            {
+                const auto& it = std::find_if(
+                    cds.get_all_sidb_locations_in_nm().cbegin(), cds.get_all_sidb_locations_in_nm().cend(),
+                    [cur_loc = sidb_nm_position<Lyt>(cds.get_phys_params(), c)](const std::pair<double, double>& loc)
+                    {
+                        return round_to_n_decimal_places(cur_loc.first, 3) == round_to_n_decimal_places(loc.first, 3) &&
+                               round_to_n_decimal_places(cur_loc.second, 3) == round_to_n_decimal_places(loc.second, 3);
```

Comment:
I must admit that I have no idea what's happening here :sweat_smile: Why do you need the nanometer positions? The SiQAD coordinates should be enough to provide grid location information.

Reply:
I reworked this in the last commit which now avoids any lookup that can trigger an exception. The problem lied in accessing the charge state based on SiQAD coordinates, since you need the index associated with the DB at a given SiQAD coordinate. I reworked it to sort a vector of indices on the order of how they would appear when printed, after this it is easy to find the index associated with the current normal cell; we simply increment the counter that indexes our indices vector each time we find a normal cell. The next commit will have a few comments to clarify this.